### PR TITLE
Kafka 3836: KStreamReduce and KTableReduce should not pass nulls to Deserializers

### PR DESCRIPTION
**Pull request by jeyhunkarimov**

### CODE REVIEW

**Comment by guozhangwang** [comment]
@jeyhunkarimov could you follow the steps to contribute to a specific JIRA in Kafka?

https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes#ContributingCodeChanges-PullRequest

And regarding to your fix, what the JIRA is targeted at is to avoid passing null to the deserializer, since null means the key does not exist in the key-value store. For example, in line 242 of `RocksDBStore` class, when `getInternal(serdes.rawKey(key))` returns null, then we could skip calling `serdes.valueFrom()` on it, which will try to deserialize the returned value but directly return null to the get call.

**Comment by jeyhunkarimov** [comment]
@guozhangwang Thank you for your information.Probably I was confused because of title of issue. I made pull request with updated version.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -239,10 +239,15 @@ public V get(K key) {
             RocksDBCacheEntry entry = cache.get(key);
 
             if (entry == null) {
-                V value = serdes.valueFrom(getInternal(serdes.rawKey(key)));
-                cache.put(key, new RocksDBCacheEntry(value));
-
-                return value;
+                byte[] byteValue  = getInternal(serdes.rawKey(key));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:242]: nit: remove extra space after `byteValue`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 2]
```diff
@@ -239,10 +239,15 @@ public V get(K key) {
...
+                if (byteValue == null){
+                    return null;
+                }
+                else{
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:246]: Coding style nit: better use `} else {` here.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 3]
```diff
@@ -239,10 +239,15 @@ public V get(K key) {
...
+                    V value = serdes.valueFrom( byteValue );
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:247]: nit: No need extra space between `(` and `)`.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 4]
```diff
@@ -239,10 +239,15 @@ public V get(K key) {
             RocksDBCacheEntry entry = cache.get(key);
 
             if (entry == null) {
-                V value = serdes.valueFrom(getInternal(serdes.rawKey(key)));
-                cache.put(key, new RocksDBCacheEntry(value));
-
-                return value;
+                byte[] byteValue  = getInternal(serdes.rawKey(key));
+                if (byteValue == null){
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:243]: nit: space `{` after `)`.

**Comment by guozhangwang** [comment]
There are a couple of other places where `StateSerdes.valueFrom` is called and we should consider checking for null values, for example:

1) `RocksDBStore` in line 255.
2) `InMemoryKeyValueLoggedStore` in line 69.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamReduce.java [hunk 5]
```diff
@@ -76,8 +76,9 @@ public void process(K key, V value) {
                 }
             }
 
-            // update the store with the new value
-            store.put(key, newAgg);
+            // update the store with the new value if the value is not null
+            if (newAgg != null)
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamReduce.java:80]: If the user defined `reduce` or `aggregate` call returns `null`, it means deleting the value for this aggregated key. So we should still put `(key, null)` into the store, which will be handled as a deletion operation.

**Pull request closed without merging**